src/Microsoft.SqlTools.ServiceLayer/Profiler/ProfilerSession.cs

```diff
@@ -1,201 +1,336 @@
+                if (events.Count > 0)
+                {
+                    lastSeenEvent = events.LastOrDefault();
+                    lastSeenId = int.Parse(lastSeenEvent.Values["event_sequence"]);
+                }
+            }
+            else    // first poll at start of session, all data is old
+            {
+                // save the last event as the beginning of the profiling session
+                if (events.Count > 0)
+                {
+                    lastSeenEvent = events.LastOrDefault();
+                    lastSeenId = int.Parse(lastSeenEvent.Values["event_sequence"]);
+                }
+
+                // ignore all events before the session began
+                events.Clear();
+            }
+        }
 
-        /// <summary>
-        /// The delay between session polls
+        /// <summary>
+        /// Indicates if the current session has completed processing and will provide no new events
         /// </summary>
-        public TimeSpan PollingDelay
-        {
-            get
-            {
-                return pollingDelay;
-            }
+        public bool Completed
+        {
+            get
+            {
+                return (sessionObserver != null) ? sessionObserver.Completed : error != null;
+            }
         }
 
-        /// <summary>
-        /// Could events have been lost in the last poll
+        private Exception error;
+        /// <summary>
+        /// Provides any fatal error encountered when processing a session
         /// </summary>
-        public bool EventsLost
-        {
-            get
-            {
-                return this.eventsLost;
-            }
+        public Exception Error
+        {
+            get
+            {
+                return sessionObserver?.Error ?? error;
+            }
+        }
+
+        /// Returns the current set of events in the session buffer.
+        /// For RingBuffer sessions, returns the content of the session ring buffer by querying the server.
+        /// For LiveTarget and LocalFile sessions, returns the events buffered in memory since the last call to GetCurrentEvents.
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerable<ProfilerEvent> GetCurrentEvents()
+        {
+            if (XEventSession == null && sessionObserver == null)
+            {
+                return Enumerable.Empty<ProfilerEvent>();
+            }
+            if (sessionObserver == null)
+            {
+                try
+                {
+                    var targetXml = XEventSession.GetTargetXml();
+
+                    XmlDocument xmlDoc = new XmlDocument();
+                    xmlDoc.LoadXml(targetXml);
+
+                    var nodes = xmlDoc.DocumentElement.GetElementsByTagName("event");
+                    var rawEvents = nodes.Cast<XmlNode>().Select(ParseProfilerEvent).ToList();
+                    FilterOldEvents(rawEvents);

```

Comment:
>     FilterOldEvents(rawEvents); [](http://example.com/codeflow?start=15&length=32)

something to consider in a future improvement - find a way to filter the events without copying them all into a list and rewriting the list. It feels like we should be able to keep some minimal state that is updated as new events are read from the target. Maybe implement an IQueryable so it can be used as a Linq expression.

Reply:
Yes, that is the next big item ticket that will be worked on (switching to live target).